File: Analysis/y_junction_fdtd/matlab/step4_phase_sweep.m
%% Y-Junction Acoustic Logic Unit — 2D FDTD Step 4
% Phase sweep demonstration.
%
% For each phase difference φ between the two inlet sources:
%     - Reset fields to zero
%     - Run to steady state (1700 steps)
%     - Measure outlet pressure over last 3 periods
%
% Outputs:
%     - Amplitude vs phase curve (3 metrics)
%     - Side-by-side pressure fields for φ = 0° and 180°

clear; clc; close all;

%% Reuse Step 3 parameters and geometry
DX = 5e-5;
C = 1500.0;
FREQ = 1.0e6;
LAMBDA = C / FREQ;
CFL = 0.5;
DT = CFL * DX / (C * sqrt(2));
coef = (C * DT / DX)^2;

LAMBDA_CELLS = round(LAMBDA / DX);
W_CHANNEL = round(0.3 * LAMBDA_CELLS);
NX = 250;
NY = 200;
L_SECTION = floor(NX / 3);
X_BEND = L_SECTION;
X_MERGE = 2 * L_SECTION;
Y_MERGE = round(NY / 2);
INLET_ANGLE_RAD = deg2rad(30);
Y_OFFSET = L_SECTION * tan(INLET_ANGLE_RAD);
Y_INLET_A = Y_MERGE + Y_OFFSET;
Y_INLET_B = Y_MERGE - Y_OFFSET;

SRC_AMP = 1.0;
STEPS_PER_PERIOD = round(1.0 / FREQ / DT);
NT = 1700;                           % ~20 periods to reach steady state
MEASURE_STEPS = 3 * STEPS_PER_PERIOD; % Last 3 periods for measurement

fprintf('Steps per period = %d\n', STEPS_PER_PERIOD);
fprintf('Total steps per run = %d\n', NT);
fprintf('Measurement window = last %d steps (%.1f periods)\n', MEASURE_STEPS, MEASURE_STEPS/STEPS_PER_PERIOD);

% Sources and probe
srcA_x = 1;
srcA_y = round(Y_INLET_A);
srcB_x = 1;
srcB_y = round(Y_INLET_B);
PROBE_X = X_MERGE + LAMBDA_CELLS;
PROBE_Y = Y_MERGE;

%% Geometry builder (copy from Step 3)
bc_mask = build_y_junction_mask(NX, NY, X_BEND, X_MERGE, Y_MERGE, W_CHANNEL, Y_INLET_A, Y_INLET_B);

%% Phase sweep
PHASE_DEGREES = 0:30:360;   % 0°, 30°, ..., 360°
PHASE_RADIANS = deg2rad(PHASE_DEGREES);

results_peak = zeros(size(PHASE_DEGREES));
results_rms = zeros(size(PHASE_DEGREES));
results_pkpk = zeros(size(PHASE_DEGREES));
snapshot_fields = struct();   % Store fields for selected phases

fprintf('\nRunning phase sweep: %d cases\n', length(PHASE_DEGREES));
fprintf('%s\n', repmat('=', 1, 50));

for idx = 1:length(PHASE_DEGREES)
    phi_deg = PHASE_DEGREES(idx);
    phi_rad = PHASE_RADIANS(idx);
    
    [probe, field] = run_fdtd(phi_rad, NT, NX, NY, srcA_x, srcA_y, srcB_x, srcB_y, PROBE_X, PROBE_Y, SRC_AMP, FREQ, DT, coef, bc_mask);
    
    % Extract measurement window (last 3 periods)
    window = probe(end-MEASURE_STEPS+1:end);
    
    % Three metrics
    peak = max(abs(window));
    rms_val = sqrt(mean(window.^2));
    pkpk = max(window) - min(window);
    
    results_peak(idx) = peak;
    results_rms(idx) = rms_val;
    results_pkpk(idx) = pkpk;
    
    % Store field for selected phases
    if ismember(phi_deg, [0, 90, 180, 270])
        field_name = sprintf('phi%d', phi_deg);
        snapshot_fields.(field_name) = field;
    end
    
    fprintf('  φ = %3d°  peak = %.4f  RMS = %.4f  Pk-Pk = %.4f\n', phi_deg, peak, rms_val, pkpk);
end
fprintf('%s\n', repmat('=', 1, 50));

%% Visualisation
scriptPath = mfilename('fullpath');
if isempty(scriptPath)
    figDir = fullfile(pwd, 'figures');
else
    [scriptDir, ~, ~] = fileparts(scriptPath);
    figDir = fullfile(scriptDir, '..', 'figures');
end
if ~exist(figDir, 'dir')
    mkdir(figDir);
end

% --- Figure 1: Amplitude vs Phase ---
figure('Position', [100 100 800 450], 'Color', 'w');
plot(PHASE_DEGREES, results_peak, 'o-', 'LineWidth', 1.5, 'DisplayName', 'Peak |p|', 'Color', [0 0.4470 0.7410]);
hold on;
plot(PHASE_DEGREES, results_rms, 's-', 'LineWidth', 1.5, 'DisplayName', 'RMS p', 'Color', [0.8500 0.3250 0.0980]);
plot(PHASE_DEGREES, results_pkpk, '^-', 'LineWidth', 1.5, 'DisplayName', 'Peak-to-peak', 'Color', [0.9290 0.6940 0.1250]);
xline(0, 'Color', [0.5 0.5 0.5], 'LineStyle', '--', 'Alpha', 0.3);
xline(180, 'Color', [0.5 0.5 0.5], 'LineStyle', '--', 'Alpha', 0.3);
ylim_vals = ylim;
text(0, ylim_vals(2)*0.95, 'Constructive\n(AND/OR)', 'HorizontalAlignment', 'center', 'FontSize', 9);
text(180, ylim_vals(2)*0.95, 'Destructive\n(XOR)', 'HorizontalAlignment', 'center', 'FontSize', 9);
xlabel('Phase difference φ [degrees]');
ylabel('Outlet pressure [Pa]');
title('Logic Gate Performance: Outlet Pressure vs Input Phase');
xticks(PHASE_DEGREES);
legend('Location', 'best');
grid on;
outPath = fullfile(figDir, 'step4_phase_sweep.png');
print(gcf, outPath, '-dpng', '-r150');
fprintf('\nSaved figure: %s\n', outPath);

% --- Figure 2: Side-by-side pressure fields ---
field_names = fieldnames(snapshot_fields);
n_fields = length(field_names);
ncols = 2;
nrows = ceil(n_fields / ncols);
figure('Position', [100 100 500*ncols 450*nrows], 'Color', 'w');

vmax = 0;
for k = 1:n_fields
    vmax = max(vmax, max(abs(snapshot_fields.(field_names{k})(:))));
end

x = linspace(0, NX*DX*1e3, NX);
y = linspace(0, NY*DX*1e3, NY);

for k = 1:n_fields
    subplot(nrows, ncols, k);
    field = snapshot_fields.(field_names{k});
    imagesc(x, y, field.');
    axis image;
    set(gca, 'YDir', 'normal');
    colormap(gca, rdBu_r(256));
    caxis([-vmax vmax]);
    colorbar;
    hold on;
    contour(x, y, bc_mask.', [0.5 0.5], 'k-', 'LineWidth', 0.5);
    % Extract phase value from field name (e.g., 'phi90' -> 90)
    phi_val = sscanf(field_names{k}, 'phi%d');
    title(sprintf('φ = %d°', phi_val));
    xlabel('x [mm]');
    ylabel('y [mm]');
end

sgtitle('Step 4: Pressure Field at Selected Phase Differences', 'FontSize', 14);
outPath = fullfile(figDir, 'step4_snapshots.png');
print(gcf, outPath, '-dpng', '-r150');
fprintf('Saved figure: %s\n', outPath);

% --- Figure 3: Logic contrast summary ---
logic_high = results_rms(1);       % φ = 0°
logic_low = results_rms(7);        % φ = 180° (index 7 = 0:30:360 -> 180 is the 7th element)
contrast = (logic_high - logic_low) / (logic_high + logic_low);

figure('Position', [100 100 500 350], 'Color', 'w');
bar([logic_high, logic_low], 'FaceColor', 'flat');
set(gca, 'XTickLabel', {'Logic "1" (φ = 0°)', 'Logic "0" (φ = 180°)'});
hold on;
yline((logic_high + logic_low)/2, 'k--', 'LineWidth', 1.5, 'Label', sprintf('Threshold = %.4f Pa', (logic_high+logic_low)/2));
ylabel('RMS pressure [Pa]');
title(sprintf('Logic Contrast Ratio = %.3f', contrast));
outPath = fullfile(figDir, 'step4_logic_contrast.png');
print(gcf, outPath, '-dpng', '-r150');
fprintf('Saved figure: %s\n', outPath);

%% Summary
fprintf('\n%s\n', repmat('=', 1, 50));
fprintf('STEP 4 COMPLETE\n');
fprintf('%s\n', repmat('=', 1, 50));
fprintf('Logic HIGH (φ = 0°):   RMS = %.4f Pa\n', logic_high);
fprintf('Logic LOW  (φ = 180°): RMS = %.4f Pa\n', logic_low);
fprintf('Contrast ratio:        %.3f\n', contrast);
fprintf('Threshold:             %.4f Pa\n', (logic_high + logic_low)/2);
fprintf('%s\n', repmat('=', 1, 50));

%% Local functions
function d = point_segment_distance(px, py, x1, y1, x2, y2)
    dx_seg = x2 - x1;
    dy_seg = y2 - y1;
    if dx_seg == 0 && dy_seg == 0
        d = hypot(px - x1, py - y1);
        return;
    end
    t = max(0.0, min(1.0, ((px - x1) * dx_seg + (py - y1) * dy_seg) / (dx_seg^2 + dy_seg^2)));
    proj_x = x1 + t * dx_seg;
    proj_y = y1 + t * dy_seg;
    d = hypot(px - proj_x, py - proj_y);
end

function mask = build_y_junction_mask(nx, ny, x_bend, x_merge, y_merge, w, y_inlet_a, y_inlet_b)
    mask = ones(nx, ny, 'int32');
    segments = {
        [0, y_inlet_a, x_bend, y_inlet_a];
        [x_bend, y_inlet_a, x_merge, y_merge];
        [0, y_inlet_b, x_bend, y_inlet_b];
        [x_bend, y_inlet_b, x_merge, y_merge];
        [x_merge, y_merge, nx - 1, y_merge];
    };
    for i = 1:nx
        for j = 1:ny
            d_min = inf;
            for s = 1:length(segments)
                seg = segments{s};
                d = point_segment_distance(i-1, j-1, seg(1), seg(2), seg(3), seg(4));
                if d < d_min
                    d_min = d;
                end
            end
            if d_min < w / 2
                mask(i, j) = 0;
            end
        end
    end
end

function lap = compute_laplacian(p, mask)
    % LEFT neighbor (i-1)
    left = zeros(size(p));
    left(2:end, :) = p(1:end-1, :);
    wall_left = false(size(mask));
    wall_left(2:end, :) = (mask(1:end-1, :) == 1);
    left(wall_left) = p(wall_left);
    neumann_left = (mask(1, :) == 1);
    left(1, neumann_left) = p(1, neumann_left);
    
    % RIGHT neighbor (i+1)
    right = zeros(size(p));
    right(1:end-1, :) = p(2:end, :);
    wall_right = false(size(mask));
    wall_right(1:end-1, :) = (mask(2:end, :) == 1);
    right(wall_right) = p(wall_right);
    neumann_right = (mask(end, :) == 1);
    right(end, neumann_right) = p(end, neumann_right);
    
    % BOTTOM neighbor (j-1)
    bottom = zeros(size(p));
    bottom(:, 2:end) = p(:, 1:end-1);
    wall_bottom = false(size(mask));
    wall_bottom(:, 2:end) = (mask(:, 1:end-1) == 1);
    bottom(wall_bottom) = p(wall_bottom);
    neumann_bottom = (mask(:, 1) == 1);
    bottom(neumann_bottom, 1) = p(neumann_bottom, 1);
    
    % TOP neighbor (j+1)
    top = zeros(size(p));
    top(:, 1:end-1) = p(:, 2:end);
    wall_top = false(size(mask));
    wall_top(:, 1:end-1) = (mask(:, 2:end) == 1);
    top(wall_top) = p(wall_top);
    neumann_top = (mask(:, end) == 1);
    top(neumann_top, end) = p(neumann_top, end);
    
    lap = left + right + bottom + top - 4 * p;
end

function [probe_trace, final_field] = run_fdtd(phase_b, NT, NX, NY, srcA_x, srcA_y, srcB_x, srcB_y, PROBE_X, PROBE_Y, SRC_AMP, FREQ, DT, coef, bc_mask)
    p_prev = zeros(NX, NY);
    p_curr = zeros(NX, NY);
    p_next = zeros(NX, NY);
    probe_trace = zeros(1, NT);
    
    for n = 0:NT-1
        t = n * DT;
        p_curr(srcA_x, srcA_y) = SRC_AMP * sin(2 * pi * FREQ * t);
        p_curr(srcB_x, srcB_y) = SRC_AMP * sin(2 * pi * FREQ * t + phase_b);
        
        laplacian = compute_laplacian(p_curr, bc_mask);
        p_next = 2 * p_curr - p_prev + coef * laplacian;
        p_next(bc_mask == 1) = p_curr(bc_mask == 1);
        
        % Swap
        temp = p_prev;
        p_prev = p_curr;
        p_curr = p_next;
        p_next = temp;
        
        probe_trace(n+1) = p_curr(PROBE_X, PROBE_Y);
    end
    
    final_field = p_curr;
end

function cmap = rdBu_r(n)
    if nargin < 1, n = 256; end
    half = floor(n/2);
    r1 = linspace(0.4039, 1, half)';
    g1 = linspace(0.0000, 1, half)';
    b1 = linspace(0.1216, 1, half)';
    r2 = linspace(1, 0.1686, n-half)';
    g2 = linspace(1, 0.2980, n-half)';
    b2 = linspace(1, 0.5373, n-half)';
    cmap = [r1 g1 b1; r2 g2 b2];
end

File: Analysis/y_junction_fdtd/step4_phase_sweep.py
"""
Y-Junction Acoustic Logic Unit — 2D FDTD Step 4
================================================

Phase sweep demonstration.

For each phase difference φ between the two inlet sources:
    - Reset fields to zero
    - Run to steady state (1700 steps)
    - Measure outlet pressure over last 3 periods

Outputs:
    - Amplitude vs phase curve (3 metrics)
    - Side-by-side pressure fields for φ = 0° and 180°
"""

import numpy as np
import matplotlib.pyplot as plt
from pathlib import Path

# ============================================================================
# Reuse Step 3 parameters and geometry
# ============================================================================
DX = 5e-5
C = 1500.0
FREQ = 1.0e6
LAMBDA = C / FREQ
CFL = 0.5
DT = CFL * DX / (C * np.sqrt(2))
coef = (C * DT / DX) ** 2

LAMBDA_CELLS = int(round(LAMBDA / DX))
W_CHANNEL = int(round(0.3 * LAMBDA_CELLS))
NX, NY = 250, 200
L_SECTION = NX // 3
X_BEND = L_SECTION
X_MERGE = 2 * L_SECTION
Y_MERGE = NY // 2
INLET_ANGLE_RAD = np.deg2rad(30)
Y_OFFSET = L_SECTION * np.tan(INLET_ANGLE_RAD)
Y_INLET_A = Y_MERGE + Y_OFFSET
Y_INLET_B = Y_MERGE - Y_OFFSET

SRC_AMP = 1.0
STEPS_PER_PERIOD = int(round(1.0 / FREQ / DT))
NT = 1700                      # ~20 periods to reach steady state
MEASURE_STEPS = 3 * STEPS_PER_PERIOD   # Last 3 periods for measurement

print(f"Steps per period = {STEPS_PER_PERIOD}")
print(f"Total steps per run = {NT}")
print(f"Measurement window = last {MEASURE_STEPS} steps ({MEASURE_STEPS/STEPS_PER_PERIOD:.1f} periods)")

# Sources and probe
srcA_x, srcA_y = 0, int(round(Y_INLET_A))
srcB_x, srcB_y = 0, int(round(Y_INLET_B))
PROBE_X = X_MERGE + LAMBDA_CELLS
PROBE_Y = Y_MERGE

# ============================================================================
# Geometry builder (copy from Step 3)
# ============================================================================

def point_segment_distance(px, py, x1, y1, x2, y2):
    dx_seg = x2 - x1
    dy_seg = y2 - y1
    if dx_seg == 0 and dy_seg == 0:
        return np.hypot(px - x1, py - y1)
    t = max(0.0, min(1.0, ((px - x1) * dx_seg + (py - y1) * dy_seg) /
                     (dx_seg**2 + dy_seg**2)))
    proj_x = x1 + t * dx_seg
    proj_y = y1 + t * dy_seg
    return np.hypot(px - proj_x, py - proj_y)


def build_y_junction_mask(nx, ny, x_bend, x_merge, y_merge, w, y_inlet_a, y_inlet_b):
    mask = np.ones((nx, ny), dtype=int)
    segments = [
        (0, y_inlet_a, x_bend, y_inlet_a),
        (x_bend, y_inlet_a, x_merge, y_merge),
        (0, y_inlet_b, x_bend, y_inlet_b),
        (x_bend, y_inlet_b, x_merge, y_merge),
        (x_merge, y_merge, nx - 1, y_merge),
    ]
    for i in range(nx):
        for j in range(ny):
            d_min = min(point_segment_distance(i, j, *seg) for seg in segments)
            if d_min < w / 2:
                mask[i, j] = 0
    return mask


bc_mask = build_y_junction_mask(NX, NY, X_BEND, X_MERGE, Y_MERGE,
                                W_CHANNEL, Y_INLET_A, Y_INLET_B)

# ============================================================================
# Laplacian with internal wall reflection (Fix 1)
# ============================================================================

def compute_laplacian(p, mask):
    left = np.zeros_like(p)
    left[1:, :] = p[:-1, :]
    wall_left = np.zeros_like(mask, dtype=bool)
    wall_left[1:, :] = (mask[:-1, :] == 1)
    left[wall_left] = p[wall_left]
    left[0, :] = np.where(mask[0, :] == 1, p[0, :], left[0, :])

    right = np.zeros_like(p)
    right[:-1, :] = p[1:, :]
    wall_right = np.zeros_like(mask, dtype=bool)
    wall_right[:-1, :] = (mask[1:, :] == 1)
    right[wall_right] = p[wall_right]
    right[-1, :] = np.where(mask[-1, :] == 1, p[-1, :], right[-1, :])

    bottom = np.zeros_like(p)
    bottom[:, 1:] = p[:, :-1]
    wall_bottom = np.zeros_like(mask, dtype=bool)
    wall_bottom[:, 1:] = (mask[:, :-1] == 1)
    bottom[wall_bottom] = p[wall_bottom]
    bottom[:, 0] = np.where(mask[:, 0] == 1, p[:, 0], bottom[:, 0])

    top = np.zeros_like(p)
    top[:, :-1] = p[:, 1:]
    wall_top = np.zeros_like(mask, dtype=bool)
    wall_top[:, :-1] = (mask[:, 1:] == 1)
    top[wall_top] = p[wall_top]
    top[:, -1] = np.where(mask[:, -1] == 1, p[:, -1], top[:, -1])

    return left + right + bottom + top - 4 * p


# ============================================================================
# Single FDTD run function
# ============================================================================

def run_fdtd(phase_b, nt, store_snapshot=False):
    """
    Run FDTD with given phase difference.
    Returns probe_trace and optionally the final pressure field.
    """
    p_prev = np.zeros((NX, NY))
    p_curr = np.zeros((NX, NY))
    p_next = np.zeros((NX, NY))
    probe_trace = []
    final_field = None

    for n in range(nt):
        t = n * DT
        p_curr[srcA_x, srcA_y] = SRC_AMP * np.sin(2 * np.pi * FREQ * t)
        p_curr[srcB_x, srcB_y] = SRC_AMP * np.sin(2 * np.pi * FREQ * t + phase_b)

        laplacian = compute_laplacian(p_curr, bc_mask)
        p_next[:] = 2 * p_curr - p_prev + coef * laplacian
        p_next[bc_mask == 1] = p_curr[bc_mask == 1]
        p_prev, p_curr, p_next = p_curr, p_next, p_prev

        probe_trace.append(p_curr[PROBE_X, PROBE_Y])

    if store_snapshot:
        final_field = p_curr.copy()

    return np.array(probe_trace), final_field


# ============================================================================
# Phase sweep
# ============================================================================
PHASE_DEGREES = np.arange(0, 361, 30)   # 0°, 30°, ..., 360°
PHASE_RADIANS = np.deg2rad(PHASE_DEGREES)

results_peak = []
results_rms = []
results_pkpk = []
snapshot_fields = {}   # Store fields for selected phases

print(f"\nRunning phase sweep: {len(PHASE_DEGREES)} cases")
print("=" * 50)

for phi_deg, phi_rad in zip(PHASE_DEGREES, PHASE_RADIANS):
    probe, field = run_fdtd(phi_rad, NT, store_snapshot=True)

    # Extract measurement window (last 3 periods)
    window = probe[-MEASURE_STEPS:]

    # Three metrics
    peak = np.max(np.abs(window))
    rms = np.sqrt(np.mean(window**2))
    pkpk = np.max(window) - np.min(window)

    results_peak.append(peak)
    results_rms.append(rms)
    results_pkpk.append(pkpk)

    # Store field for selected phases
    if phi_deg in [0, 90, 180, 270]:
        snapshot_fields[phi_deg] = field

    print(f"  φ = {phi_deg:3d}°  peak = {peak:.4f}  RMS = {rms:.4f}  Pk-Pk = {pkpk:.4f}")

print("=" * 50)

# Convert to arrays
results_peak = np.array(results_peak)
results_rms = np.array(results_rms)
results_pkpk = np.array(results_pkpk)

# ============================================================================
# Visualisation
# ============================================================================
fig_dir = Path(__file__).parent / "figures"
fig_dir.mkdir(exist_ok=True)

# --- Figure 1: Amplitude vs Phase ---
fig, ax = plt.subplots(figsize=(10, 5))

ax.plot(PHASE_DEGREES, results_peak, 'o-', label='Peak |p|', color='C0')
ax.plot(PHASE_DEGREES, results_rms, 's-', label='RMS p', color='C1')
ax.plot(PHASE_DEGREES, results_pkpk, '^-', label='Peak-to-peak', color='C2')

ax.axvline(0, color='gray', linestyle='--', alpha=0.3)
ax.axvline(180, color='gray', linestyle='--', alpha=0.3)
ax.text(0, ax.get_ylim()[1]*0.95, 'Constructive\n(AND/OR)', ha='center', fontsize=9)
ax.text(180, ax.get_ylim()[1]*0.95, 'Destructive\n(XOR)', ha='center', fontsize=9)

ax.set_xlabel('Phase difference φ [degrees]')
ax.set_ylabel('Outlet pressure [Pa]')
ax.set_title('Logic Gate Performance: Outlet Pressure vs Input Phase')
ax.set_xticks(PHASE_DEGREES)
ax.legend()
ax.grid(True)
plt.tight_layout()
out_path = fig_dir / 'step4_phase_sweep.png'
plt.savefig(out_path, dpi=150)
print(f"\nSaved figure: {out_path}")
plt.close()

# --- Figure 2: Side-by-side pressure fields ---
fig, axes = plt.subplots(2, 2, figsize=(12, 10))
vmax = max(np.abs(f).max() for f in snapshot_fields.values())

for ax, (phi_deg, field) in zip(axes.flatten(), snapshot_fields.items()):
    im = ax.imshow(field.T, origin='lower', cmap='RdBu_r',
                   vmin=-vmax, vmax=vmax,
                   extent=[0, NX*DX*1e3, 0, NY*DX*1e3])
    ax.contour(bc_mask.T, levels=[0.5], colors='k', linewidths=0.5,
               extent=[0, NX*DX*1e3, 0, NY*DX*1e3])
    ax.set_title(f'φ = {phi_deg}°')
    ax.set_xlabel('x [mm]')
    ax.set_ylabel('y [mm]')
    fig.colorbar(im, ax=ax, fraction=0.046)

plt.suptitle('Step 4: Pressure Field at Selected Phase Differences', fontsize=14)
plt.tight_layout()
out_path = fig_dir / 'step4_snapshots.png'
plt.savefig(out_path, dpi=150)
print(f"Saved figure: {out_path}")
plt.close()

# --- Figure 3: Logic contrast summary ---
logic_high = results_rms[0]      # φ = 0°
logic_low = results_rms[6]       # φ = 180°
contrast = (logic_high - logic_low) / (logic_high + logic_low)

fig, ax = plt.subplots(figsize=(6, 4))
ax.bar(['Logic "1"\n(φ = 0°)', 'Logic "0"\n(φ = 180°)'],
       [logic_high, logic_low],
       color=['C0', 'C3'])
ax.axhline((logic_high + logic_low)/2, color='k', linestyle='--',
           label=f'Threshold = {(logic_high+logic_low)/2:.4f} Pa')
ax.set_ylabel('RMS pressure [Pa]')
ax.set_title(f'Logic Contrast Ratio = {contrast:.3f}')
ax.legend()
plt.tight_layout()
out_path = fig_dir / 'step4_logic_contrast.png'
plt.savefig(out_path, dpi=150)
print(f"Saved figure: {out_path}")
plt.close()

# ============================================================================
# Summary
# ============================================================================
print("\n" + "=" * 50)
print("STEP 4 COMPLETE")
print("=" * 50)
print(f"Logic HIGH (φ = 0°):   RMS = {logic_high:.4f} Pa")
print(f"Logic LOW  (φ = 180°): RMS = {logic_low:.4f} Pa")
print(f"Contrast ratio:        {contrast:.3f}")
print(f"Threshold:             {(logic_high + logic_low)/2:.4f} Pa")
print("=" * 50)
